platform/linux-generic/include/odp_llqueue.h

```diff
@@ -41,6 +41,7 @@ static odp_bool_t llq_on_queue(struct llnode *node);
  *****************************************************************************/
 
 #define SENTINEL ((void *)~(uintptr_t)0)
+#define MAX_SPIN_COUNT 100000
```

Comment:
That long a wait looks a bit excessive to me. I wonder if something like 1000 or even less be better.

Reply:
The number I set is about 5M cpu cycles in one of our boxes, roughly 0.25s. With this number, I find most enqueue will just finish without going to restart. 

My goal is to not restart for normal wait, but only for the possible case of endless loop.

Reply:
5 M cycles or 0.25 seconds is a very long time to stop a thread that may have possibly latency sensitive work to do. The problem of an indefinite wait is replaced with the problem of a long wait that can easily disturb many applications and cause packet loss and whatnot.

Wait time longer than necessary does not provide any benefit but makes dequeue last longer if the simultaneous dequeue occurs in a bad way. "Too short" a wait time merely causes a extra load-linked accesses.

If we want to avoid extra load-linked operations when enqueue happens simultaneously with the dequeue of the previous node, the wait needs to be only as long as it takes for the new next pointer to become visible after the sc was done. I would think it does not typically take a huge number of cycles.

Reply:
Now that I look at the existing code, I think there may be a problem with memory ordering. llq_enqueue() calls scd() with release memory order, but the llq_dequeue() calls lld() with relaxed memory order instead of acquire which would pair with the release. Because of this, it seems that there is no guarantee that the writing of null to node->next by llq_enqueue() becomes visible in llq_dequeue() in time. So the llq_dequeue() may see a stale non-null value in the next pointer.

If this is indeed the case, then I think another patch is needed (it is best to keep different fixes separate).

Reply:
My goal indeed is to avoid extra load-linked operations when enqueue happens simultaneously with the dequeue of the previous node.

In my box, I find that it can take some time for dequeue to wait (by checking if goto happens). And that is why I come up with the number 100000. Of course, this is only from one profile of my box.

If you insist, I can change it to some smaller value like 1000. The one I used in our code is 100000.

Reply:
> Now that I look at the existing code, I think there may be a problem with memory ordering. llq_enqueue() calls scd() with release memory order, but the llq_dequeue() calls lld() with relaxed memory order instead of acquire which would pair with the release. Because of this, it seems that there is no guarantee that the writing of null to node->next by llq_enqueue() becomes visible in llq_dequeue() in time. So the llq_dequeue() may see a stale non-null value in the next pointer.
> 
> If this is indeed the case, then I think another patch is needed (it is best to keep different fixes separate).

Maybe this also contributes to the long cycles that dequeue needs to wait?

Reply:
> 5 M cycles or 0.25 seconds is a very long time to stop a thread that may have possibly latency sensitive work to do. The problem of an indefinite wait is replaced with the problem of a long wait that can easily disturb many applications and cause packet loss and whatnot.
> 
> Wait time longer than necessary does not provide any benefit but makes dequeue last longer if the simultaneous dequeue occurs in a bad way. "Too short" a wait time merely causes a extra load-linked accesses.
> 
> If we want to avoid extra load-linked operations when enqueue happens simultaneously with the dequeue of the previous node, the wait needs to be only as long as it takes for the new next pointer to become visible after the sc was done. I would think it does not typically take a huge number of cycles.

---

I would think it should not take too many cycles too. But in my box, I find that it can take some time for dequeue to wait (by checking if goto happens). And that is why I come up with the number 100000. Of course, this is only from one profile of my box.

And my goal indeed is to avoid extra load-linked operations when enqueue happens simultaneously with the dequeue of the previous node. Too many retries also wast cpu cycles during spinning, right?

Anyway if you insist, I can change it to some smaller value like 1000. The one I used in my box is 100000.

Reply:
set to 1000.